Userland/Libraries/LibJS/Runtime/IteratorOperations.cpp

```diff
@@ -84,53 +113,60 @@ ThrowCompletionOr<Value> iterator_value(GlobalObject& global_object, Object& ite
 }
 
 // 7.4.5 IteratorStep ( iteratorRecord ), https://tc39.es/ecma262/#sec-iteratorstep
-ThrowCompletionOr<Object*> iterator_step(GlobalObject& global_object, Object& iterator)
+ThrowCompletionOr<Object*> iterator_step(GlobalObject& global_object, Iterator const& iterator_record)
 {
-    auto* result = TRY(iterator_next(iterator));
+    // 1. Let result be ? IteratorNext(iteratorRecord).
+    auto* result = TRY(iterator_next(global_object, iterator_record));
+
+    // 2. Let done be ? IteratorComplete(result).
     auto done = TRY(iterator_complete(global_object, *result));
 
+    // 3. If done is true, return false.
     if (done)
         return nullptr;
 
+    // 4. Return result.
     return result;
 }
 
 // 7.4.6 IteratorClose ( iteratorRecord, completion ), https://tc39.es/ecma262/#sec-iteratorclose
 // 7.4.8 AsyncIteratorClose ( iteratorRecord, completion ), https://tc39.es/ecma262/#sec-asynciteratorclose
 // NOTE: These only differ in that async awaits the inner value after the call.
-static Completion iterator_close_(Object& iterator, Completion completion, IteratorHint iterator_hint)
+static Completion iterator_close_impl(GlobalObject& global_object, Iterator const& iterator_record, Completion completion, IteratorHint iterator_hint)
 {
-    auto& vm = iterator.vm();
-    auto& global_object = iterator.global_object();
+    auto& vm = global_object.vm();
+
+    // 1. Assert: Type(iteratorRecord.[[Iterator]]) is Object.
+
+    // 2. Let iterator be iteratorRecord.[[Iterator]].
+    auto* iterator = iterator_record.iterator;
 
     // The callers of iterator_close() are often in an exceptional state.
     // Temporarily clear that exception for invocation(s) to Call.
     TemporaryClearException clear_exception(vm);
 
     // 3. Let innerResult be GetMethod(iterator, "return").
-    auto inner_result_or_error = Value(&iterator).get_method(global_object, vm.names.return_);
-    Value inner_result;
+    auto inner_result = ThrowCompletionOr<Value> { js_undefined() };
+    auto get_method_result = Value(iterator).get_method(global_object, vm.names.return_);
+    if (get_method_result.is_error())
+        inner_result = get_method_result.release_error();
```

Comment:
~Am I misreading this or are you only assigning to `inner_result` in the error case?
Will the inner result now always be `undefined` or an error?~

Ok, it's a bit more tricky than I thought.

Or perhaps should the check in the if of step 4. be about `get_method_result`?
(Oh I guess this could work, but it isn't very clear to me at the moment so maybe at least adding a comment could help clear things up?)

Reply:
> Or perhaps should the check in the if of step 4. be about get_method_result?

Nope, we might just have moved the error out of the optional in the line above, in which case it would crash. `ThrowCompletionOr<T>` -> `ThrowCompletionOr<Value>` isn't super comfortable right now, but luckily quite rare.

Reply:
Ah right, yeah it's just confusing that inner_result is `js_undefined()` unless something failed but I get it now I guess no comment needed anymore

Reply:
> we might just have moved the error

...and when that happens, the `!is_error()` check is true, even though we don't have a value.